File: JAVA_code/Lecture_32.java
                                        //Lecture-32
                                        //August-2

import java.util.*;
public class Lecture_32
{   
    
    //1.SHARES_PROFIT_PROBLEM       (to get max profit by buying and selling shares
                                    //without holding more than 1 share at a time).
    static int sharesProfit(int[] arr, int k) {
        int[][] dp = new int[2][arr.length];

        dp[0][0] = -1 * arr[0];         //buying the first stock.
        dp[1][0] = 0;                   //nothing to buy.

        for(int j = 1; j < dp[0].length; j++) {
            for(int i = 0; i < dp.length; i++) {
                if(i == 0) {
                    dp[i][j] = Math.max(dp[i][j - 1], dp[i + 1][j - 1] - arr[j]);
                }
                else {
                    dp[i][j] = Math.max(dp[i][j - 1], dp[i - 1][j - 1] + arr[j] - k);
                }
            }
        }

        return dp[dp.length - 1][dp[0].length - 1];
    }

    //REVIEWED.












    //2.LONGEST_COMMON_SUBSEQUENCE      (just to find max length of LCS)

    static int[][] longestCommonSubSequence(String s1,String s2)                    //TO GET LENGTH LONGEST COMMON SUBSEQUENCE.
    {
        int[][] global=new int[s1.length()+1][s2.length()+1];

        for(int r=global.length-1;r>=0;r--)
        {
            for(int c=global[0].length-1;c>=0;c--)
            {
                if(r==global.length-1 || c==global[0].length-1)
                {
                    global[r][c]=0;
                }
                else
                {
                    if(s1.charAt(r)==s2.charAt(c))      //if characters are same.
                        global[r][c]=1+global[r+1][c+1];

                    else                                //if characters are not same.
                        global[r][c]=Math.max(global[r+1][c],global[r][c+1]);
                }
            }
        }
        return global;
    }

    //REVIEWED.












    //3.PRINTING_THE_LCS        (it uses the above dp creating function)

    static void longestCommonSubSequencePath(String s1,String s2,int i,int j,int[][] global2,String ans)   //TO GET LONGEST COMMON
    {                                                                                                      //SUBSEQUENCE.
        if(i==global2.length-1 && j==global2.length-1)
        {
            System.out.println(ans);
            return;
        }

        if(s1.charAt(i)==s2.charAt(j))
        {
            ans+=s1.charAt(i);
            longestCommonSubSequencePath(s1, s2, i+1, j+1, global2, ans);
        }
        else
        {
            if(global2[i+1][j]>global2[i][j+1])
            {
                longestCommonSubSequencePath(s1, s2, i+1, j, global2, ans);
            }
            else
            {
                longestCommonSubSequencePath(s1, s2, i, j+1, global2, ans);
            }
        }
    }                                                                                 //REVIEWED.

    //REVEIWED.













    //4.LONGEST_PALINDROMIC_SUBSEQUENCE         (to get the length of LPS)

    static int[][] longestPalindromeSubSequence(String s)
    {
        int[][] global3=new int[s.length()][s.length()];

        for(int gap=0;gap<global3.length;gap++)                  //diagonal loop, (for a matrix).
         {   
            for(int r=0;r<global3.length-gap;r++)
            {   
                int c=r+gap;

                if(gap==0)
                {
                    global3[r][c]=1;
                }
                else if(gap==1)
                {
                    if(s.charAt(r)==s.charAt(c))
                        global3[r][c]=2;
                    else
                        global3[r][c]=1;
                }
                else if(s.charAt(r)==s.charAt(c))
                {
                    global3[r][c]=2+global3[r+1][c-1];
                }
                else
                    global3[r][c]=Math.max(global3[r+1][c],global3[r][c-1]);
            
            }
        }

        return global3;
    }                                                                       //REVIEWED.

    //REVIEWED.













    public static void main(String[] args)
    {
        
        //1.SHARE_PROFIT_PROBLEM        (to get max profit, from shares)

        /*
        int[] arr = {9,1,3,10,1,4,6,9};
        int k = 2;

        System.out.println("The max profit from shares is " + sharesProfit(arr, k));
        */

        //REVIEWED.



        





        //2. LONGEST_COMMON_SUBSEQUENCE         (only the length of LCS)

        /*
        int[][] global= longestCommonSubSequence("abcd","aebd");               //count of longest common subsequence.
        
        System.out.println("The length of LCS is " + global[0][0]);
        */

        //reviewed.
    
    







        //3.Printing_the_LCS        (this function uses DP of LCS)

        /*
        int[][] global2=longestCommonSubSequence("abcd","aebd");                //to get longest common subsequence, by recursion
        longestCommonSubSequencePath("abcd","aebd",0,0,global2,"");             //and DP both.
        */

        //reviewed.









        //4.LONGEST_PALINDROMIC_SUBSEQUENCE         (only the length of LPS)

        /*
        int[][] global3=longestPalindromeSubSequence("abckycbc");                //length of longest palindromic sub sequence.
        
        System.out.println("The length of LPS is " + global3[0][global3[0].length - 1]);
        */

        //reviewed.
    
    
    
    }
}

File: C++_code/Lecture_32.cpp
                                //Lecture-32
                                //August-2

#include<iostream>
using namespace std;



//1.SHARES_PROFIT_PROBLEM       (to get max profit by buying and selling shares,
                                //without holding more than 1 share, at a time.
int sharesProfit(int arr[], int n, int k) {
    int dp[2][n];
    
    dp[0][0] = -1*arr[0];
    dp[1][0] = 0;

    for(int j = 1; j < n; j++) {
        for(int i = 0; i < 2; i++) {
            dp[i][j] = 0;
            if(i == 0) {                //buying row.
                dp[i][j] = max(dp[i][j - 1], dp[i + 1][j - 1] - arr[j]);
            }
            else {                      //selling row.
                dp[i][j] = max(dp[i][j - 1], dp[i - 1][j - 1] + arr[j] - k);
            }
        }
    }
    return dp[1][n - 1];
}

//REVIEWED.













//2.LONGEST_COMMON_SUBSEQUENCE          (just to find the maximum length of LCS)

int lengthOfLCS(string s1, string s2) {
    int dp[s1.length() + 1][s2.length() + 1];

    for(int i = s1.length(); i >= 0; i--) {
        for(int j = s2.length(); j >= 0; j--) {
            if(i == s1.length() || j == s2.length()) {      //(anyone is empty subsequence)
                dp[i][j] = 0;
            }
            else if(s1.at(i) == s2.at(j)) {             //(row and col character matches)
                dp[i][j] = 1 + dp[i + 1][j + 1];
            }
            else {                              //(row and col charater doesn't matches)
                dp[i][j] = max(dp[i + 1][j], dp[i][j + 1]);
            }
        }
    } 

    return dp[0][0];
}

//REVIEWED.













//3. PRINTING_THE_LCS           (printing the actual LCS)

void getLCS(int dp[][5], int i, int j, string psf, string s1, string s2) {      //recursive function.
    if(i == s1.length() && j == s2.length()) {
        cout<<"The actual LCS is "<<psf;
        return;
    }

    if(s1.at(i) == s2.at(j)) {
        getLCS(dp, i + 1, j + 1, psf + s1.at(i), s1, s2);
    }
    else {
        if(dp[i + 1][j] > dp[i][j + 1]) {
            getLCS(dp, i + 1, j, psf, s1, s1);
        }
        else {
            getLCS(dp, i, j + 1, psf, s1, s2);
        }
    }
}



void printLCS(string s1, string s2) {       //(dp function)
    
    // int dp[s1.length() + 1][s2.length() + 1];        //(this is also correct, but...)
    
    int dp[5][5];       //as we have to pass it to another function, so const size.

    for(int i = s1.length(); i >= 0; i--) {
        for(int j = s2.length(); j >= 0; j--) {
            if(i == s1.length() || j == s2.length()) {
                dp[i][j] = 0;
            }
            else if(s1.at(i) == s2.at(j)) {
                dp[i][j] = 1 + dp[i + 1][j + 1];
            }
            else {
                dp[i][j] = max(dp[i + 1][j], dp[i][j + 1]);
            }
        }
    }
    getLCS(dp, 0, 0, "", s1, s2);
}

//REVIEWED.














//4.LONGEST_PALINDROMIC_SUBSEQUENCE         (to get the length of LPS only)

int lengthOfLPS(string s) {
    int dp[s.length()][s.length()];

    for(int l = 1; l <= s.length(); l++) {
        for(int i = 0, j = l - 1; i < s.length() && j < s.length(); i++, j++) {
            if(l == 1) {
                dp[i][j] = 1;
            }
            else if(l == 2) {
                if(s.at(i) == s.at(j)) {
                    dp[i][j] = 2;
                }
                else {
                    dp[i][j] = 1;
                }
            }
            else {
                if(s.at(i) == s.at(j)) {
                    dp[i][j] = 2 + dp[i + 1][j - 1];
                }
                else {
                    dp[i][j] = max(dp[i][j - 1], dp[i + 1][j]);
                }
            }
        }
    }

    return dp[0][s.length() - 1];
}

//REVIEWED.










int main()
{
    
    //1.SHARE_PROFIT_PROBLEM       (to get max profit, without holding more than one share at a time.)

    /*   
    int arr[] = {9,1,3,10,1,4,6,9};
    int k = 2;

    cout<<"The max Profit from shares is "<<sharesProfit(arr, 8, k);
    */

    //reviewed.









    //2. LONGEST_COMMON_SUBSEQUENCE     (only the length of LCS)

    /*
    string s1 = "abcd";
    string s2 = "aebd";

    cout<<"The length of LCS is"<<lengthOfLCS(s1, s2);
    */

    //reviewed.










    //3. Printing_the_LCS

    /*
    string s1 = "abcd";
    string s2 = "aebd";

    printLCS(s1, s2);
    */

    //reviewed.









    //4.LONGEST_PALINDROMIC_SUBSEQUENCE        (to find the length of LPS)

    /*
    string s = "abckycbc";
    cout<<"The length of LPS is "<<lengthOfLPS(s);
    */

   //reviewed.


}